File: scope104/src/main/java/com/petro/scope104/ui/select/SelectRVAdapter.java
package com.petro.scope104.ui.select;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.petro.scope104.databinding.ItemSelectBinding;

public class SelectRVAdapter extends ListAdapter<SelectItem, SelectRVAdapter.SelectViewHolder> {

    protected SelectRVAdapter() {
        super(new DiffUtil.ItemCallback<SelectItem>() {
            @Override
            public boolean areItemsTheSame(@NonNull SelectItem oldItem, @NonNull SelectItem newItem) {
                return oldItem.getTitle().equals(newItem.getTitle());
            }

            @Override
            public boolean areContentsTheSame(@NonNull SelectItem oldItem, @NonNull SelectItem newItem) {
                return oldItem.getTitle().equals(newItem.getTitle()) && (oldItem.getValue() == newItem.getValue());
            }
        });
    }

    @NonNull
    @Override
    public SelectViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new SelectViewHolder(ItemSelectBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull SelectViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    class SelectViewHolder extends RecyclerView.ViewHolder {
        private final ItemSelectBinding binding;

        public SelectViewHolder(ItemSelectBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        boolean isChangingCheckedState = false;
        public void bind(SelectItem selectItem) {
            binding.getRoot().setText(selectItem.getTitle());

            isChangingCheckedState = true;
            // Do not notify the CheckedChangeListener here
            binding.getRoot().setChecked(selectItem.getValue());
            isChangingCheckedState = false;

            binding.getRoot().setOnCheckedChangeListener((buttonView, isChecked) -> {
                if (isChangingCheckedState) return;
                selectItem.setValue(isChecked);
                itemView.post(() -> notifyItemChanged(getAdapterPosition()));
            });
        }
    }
}

File: scope104/src/main/kotlin/com/petro/scope104/presentation/select/SelectRVAdapter.kt
package com.petro.scope104.presentation.select

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.CompoundButton
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.petro.scope104.databinding.ItemSelectBinding
import com.petro.scope104.presentation.select.SelectRVAdapter.SelectViewHolder

class SelectRVAdapter :
    ListAdapter<SelectItem, SelectViewHolder>(object : DiffUtil.ItemCallback<SelectItem>() {
        override fun areItemsTheSame(oldItem: SelectItem, newItem: SelectItem): Boolean {
            return oldItem.title == newItem.title
        }

        override fun areContentsTheSame(oldItem: SelectItem, newItem: SelectItem): Boolean {
            return oldItem.title == newItem.title && oldItem.value == newItem.value
        }
    }) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SelectViewHolder {
        return SelectViewHolder(
            ItemSelectBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: SelectViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class SelectViewHolder(private val binding: ItemSelectBinding) : RecyclerView.ViewHolder(
        binding.root
    ) {
        private var isChangingCheckedState = false
        fun bind(selectItem: SelectItem) {
            binding.root.text = selectItem.title
            isChangingCheckedState = true
            // Do not notify the CheckedChangeListener here
            binding.root.isChecked = selectItem.value
            isChangingCheckedState = false
            binding.root.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                if (isChangingCheckedState) return@setOnCheckedChangeListener
                selectItem.value = isChecked
                itemView.post { notifyItemChanged(adapterPosition) }
            }
        }
    }
}